query-engine/connectors/sql-query-connector/src/query_builder/read.rs

```diff
@@ -67,10 +68,70 @@ where
     columns.fold(query.into_select(model), |acc, col| acc.column(col))
 }
 
-pub fn count_by_model(model: &ModelRef, query_arguments: QueryArguments) -> Select<'static> {
-    let selected_columns = model.primary_identifier().as_columns();
-    let base_query = get_records(model, selected_columns, query_arguments);
-    let table = Table::from(base_query).alias("sub");
+/// Generates a query of the form:
+/// ```sql
+/// SELECT
+///     COUNT(*),
+///     SUM(`float`),
+///     SUM(`int`),
+///     AVG(`float`),
+///     AVG(`int`),
+///     MIN(`float`),
+///     MIN(`int`),
+///     MAX(`float`),
+///     MAX(`int`)
+/// FROM
+///     (
+///         SELECT
+///             `Table`.`id`,
+///             `Table`.`float`,
+///             `Table`.`int`
+///         FROM
+///             `Table`
+///         WHERE
+///             1 = 1
+///     ) AS `sub`;
+/// ```
+pub fn aggregate(model: &ModelRef, aggregators: &[Aggregator], args: QueryArguments) -> Select<'static> {
+    let columns = extract_columns(model, &aggregators);
+    let sub_query = get_records(model, columns.into_iter(), args);
+    let sub_table = Table::from(sub_query).alias("sub");
 
-    Select::from_table(table).value(count(asterisk()))
+    aggregators
+        .into_iter()
+        .fold(Select::from_table(sub_table), |select, next_op| match next_op {
+            Aggregator::Count => select.value(count(asterisk())),
+
+            Aggregator::Average(fields) => fields
+                .into_iter()
+                .fold(select, |select, next_field| select.value(avg(next_field.name.clone()))),
+
+            Aggregator::Sum(fields) => fields
+                .into_iter()
+                .fold(select, |select, next_field| select.value(sum(next_field.name.clone()))),
+
+            Aggregator::Min(fields) => fields
+                .into_iter()
+                .fold(select, |select, next_field| select.value(min(next_field.name.clone()))),
+
+            Aggregator::Max(fields) => fields
+                .into_iter()
+                .fold(select, |select, next_field| select.value(max(next_field.name.clone()))),
+        })
+}
+
+fn extract_columns(model: &ModelRef, aggregators: &[Aggregator]) -> Vec<Column<'static>> {
```

Comment:
maybe `Aggregator::extract_columns(model: &ModelRef)`?

Reply:
No, this leaks SQL specifics into the connector crate.